firebase-firestore/src/main/java/com/google/firebase/firestore/core/TransactionRunner.java

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.core;
+
+import androidx.annotation.NonNull;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.common.base.Function;
+import com.google.firebase.firestore.FirebaseFirestoreException;
+import com.google.firebase.firestore.remote.Datastore;
+import com.google.firebase.firestore.remote.RemoteStore;
+import com.google.firebase.firestore.util.AsyncQueue;
+import com.google.firebase.firestore.util.AsyncQueue.TimerId;
+import com.google.firebase.firestore.util.ExponentialBackoff;
+
+/**
+ * TransactionRunner encapsulates the logic needed to run and retry transactions so that the caller
+ * does not have to manage the backoff and retry count through recursive calls.
+ */
+public class TransactionRunner<TResult> {
+  private static final int RETRY_COUNT = 5;
+  private AsyncQueue asyncQueue;
+  private RemoteStore remoteStore;
+  private Function<Transaction, Task<TResult>> updateFunction;
+  private int retriesLeft;
+
+  private ExponentialBackoff backoff;
+  private TaskCompletionSource<TResult> taskSource = new TaskCompletionSource<>();
+
+  public TransactionRunner(
+      AsyncQueue asyncQueue,
+      RemoteStore remoteStore,
+      Function<Transaction, Task<TResult>> updateFunction) {
+
+    this.asyncQueue = asyncQueue;
+    this.remoteStore = remoteStore;
+    this.updateFunction = updateFunction;
+    this.retriesLeft = RETRY_COUNT;
+
+    backoff = new ExponentialBackoff(asyncQueue, TimerId.RETRY_TRANSACTION);
+  }
+
+  /** Runs the transaction and sets the result in taskSource. */
+  public Task<TResult> run() {
+    runWithBackoff();
+    return taskSource.getTask();
+  }
+
+  private void runWithBackoff() {
+    backoff.backoffAndRun(
+        () -> {
+          final Transaction transaction = remoteStore.createTransaction();
+          updateFunction
+              .apply(transaction)
+              .addOnCompleteListener(
+                  asyncQueue.getExecutor(),
+                  new OnCompleteListener<TResult>() {
```

Comment:
Sorry for not mentioning before, but you can replace these anonymous classes with lambda syntax, which looks a bit cleaner:

```
    backoff.backoffAndRun(
        () -> {
          final Transaction transaction = remoteStore.createTransaction();
          updateFunction
              .apply(transaction)
              .addOnCompleteListener(
                  asyncQueue.getExecutor(),
                  (@NonNull Task<TResult> userTask) -> {
                    if (!userTask.isSuccessful()) {
                      handleTransactionError(userTask);
                    } else {
                      transaction
                          .commit()
                          .addOnCompleteListener(
                              asyncQueue.getExecutor(),
                              (@NonNull Task<Void> commitTask) -> {
                                if (commitTask.isSuccessful()) {
                                  taskSource.setResult(userTask.getResult());
                                } else {
                                  handleTransactionError(commitTask);
                                }
                              });
                    }
                  });
        });
```